internal/venice-client-common/src/main/java/com/linkedin/venice/utils/concurrent/CloseableThreadLocal.java

```diff
@@ -0,0 +1,72 @@
+package com.linkedin.venice.utils.concurrent;
+
+import com.linkedin.venice.utils.HashCodeComparator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * A wrapper of {@link ThreadLocal} for {@link AutoCloseable} objects.
+ * Java's {@link ThreadLocal} only dereferences the thread-local objects
+ * when the {@link Thread} exits which triggers their garbage collection.
+ * It does not {@code close} the {@link AutoCloseable} thread-local objects.
+ * @param <T> The {@link AutoCloseable} type whose objects will be held by an object of this class.
+ */
+public class CloseableThreadLocal<T extends AutoCloseable> implements AutoCloseable {
+  private static final Logger LOGGER = LogManager.getLogger(CloseableThreadLocal.class);
```

Comment:
Shall we throw `UnsupportedOperationException` for `set` method? Otherwise, the replaced internal variable won't be `closed` properly?

Reply:
This class doesn't extend `ThreadLocal` at the moment. It just creates an internal `ThreadLocal` delegate. Should it extend `ThreadLocal`?

Reply:
My bad, that should be fine then.
We don't need to extend `ThreadLocal` here.